lib/spack/spack/ci.py

```diff
@@ -930,13 +930,13 @@ def generate_gitlab_ci_yaml(
     except bindist.FetchCacheError as e:
         tty.warn(e)
 
-    spec_labels, dependencies, stages = stage_spec_jobs(
-        [
-            concrete
-            for abstract, concrete in env.concretized_specs()
-            if abstract in env.spec_lists["specs"]
-        ]
-    )
+    concretized_specs = {
+        concrete: abstract
+        for abstract, concrete in env.concretized_specs()
```

Comment:
Unfortunately, this line only contains abstract specs for the top level specs. for instance, given: `perl ^berkeley-db~docs`, `perl` would be the only entry in this dict.

It would be good to get the abstract spec for the dependency spec, but I'm not sure of a good way of programmatically grab this..